# SMU Master of Science (Economics) Programming Workshop in Python


## Introduction
In today's class, we will be learning more about Data Cleaning and Manipulation with pandas and NumPy. Both pandas and NumPy are great libraries that contain many useful functions for data cleaning and manipulation. 

---

pandas
![](images/pandas_logo.png "pandas")
pandas is an open source, BSD-licensed library providing high-performance, easy-to-use data structures and data analysis tools for the Python programming language.

---

NumPy
![](images/numpy_logo.png "NumPy")
NumPy is a library for the Python programming language, adding support for large, multi-dimensional arrays and matrices, along with a large collection of high-level mathematical functions to operate on these arrays.

You can find more information about pandas [here](https://pandas.pydata.org/), and more information about NumPy [here](https://www.numpy.org/).


Before we go dive deep into data cleaning and manipulation, it is useful to take a step back and think about the differences between a function and a method:

1. A function is a "global" function that can be used on different kinds of data structures. For example, the `len` function can be used on lists and strings.
2. A method is a "local" function that can only be used on a specific kind of data structure. For example, the method `pivot_table` can only be used on the DataFrame data structure (we will see this in more detail).

#### Case Study Approach

To learn how we can use pandas and NumPy, we will adopt a case study approach. We have attached 2 datasets (which you can find in the /data folder):

1. The first dataset is from [Kaggle](https://www.kaggle.com/), a machine learning platform. More information about the data can be found [here](https://www.kaggle.com/c/titanic/data). This dataset contains information about the passengers that were involved in Titanic's sinking in 1912.

2. The second dataset is a dataset on movie ratings that comes from [GroupLens](https://grouplens.org/). In this dataset, there are 1,000,000 entries on movie ratings by users. The metadata can be found [here](http://files.grouplens.org/datasets/movielens/ml-1m-README.txt). 

---

We will begin our data analysis on the Titanic dataset first. We begin by importing key libraries, such as pandas and NumPy library into Jupyter, by calling the "import" function. In this case, we will only be dealing with the training data.

---

#### Data Manipulation with Titanic
In the first part of data manipulation, we learn how to get summary statistics of the dataset, and the types of data we are dealing with.

if data is the filename,

1. data.shape returns the dimension of the dataset we are dealing with
2. data.head() returns the first five rows of the dataset
3. data.describe() returns the summary statistics of the dataset

In [1]:
import pandas as pd
import numpy as np

# Since data is in the form of a csv file, we can use pd.read_csv
train_data = pd.read_csv('data/titanic/train.csv', index_col = 0) # The first column is the index 
print(train_data.shape)
# train_data.head()
train_data.describe(include='all') # Describe gives a summary of the dataframe 

(891, 11)


Survived      Pclass                  Name   Sex         Age  \
count   891.000000  891.000000                   891   891  714.000000   
unique         NaN         NaN                   891     2         NaN   
top            NaN         NaN  Waelens, Mr. Achille  male         NaN   
freq           NaN         NaN                     1   577         NaN   
mean      0.383838    2.308642                   NaN   NaN   29.699118   
std       0.486592    0.836071                   NaN   NaN   14.526497   
min       0.000000    1.000000                   NaN   NaN    0.420000   
25%       0.000000    2.000000                   NaN   NaN   20.125000   
50%       0.000000    3.000000                   NaN   NaN   28.000000   
75%       1.000000    3.000000                   NaN   NaN   38.000000   
max       1.000000    3.000000                   NaN   NaN   80.000000   

             SibSp       Parch Ticket        Fare        Cabin Embarked  
count   891.000000  891.000000    891  891.000000          204      889  
unique         NaN         NaN    681         NaN          147        3  
top            NaN         NaN   1601         NaN  C23 C25 C27        S  
freq           NaN         NaN      7         NaN            4      644  
mean      0.523008    0.381594    NaN   32.204208          NaN      NaN  
std       1.102743    0.806057    NaN   49.693429          NaN      NaN  
min       0.000000    0.000000    NaN    0.000000          NaN      NaN  
25%       0.000000    0.000000    NaN    7.910400          NaN      NaN  
50%       0.000000    0.000000    NaN   14.454200          NaN      NaN  
75%       1.000000    0.000000    NaN   31.000000          NaN      NaN  
max       8.000000    6.000000    NaN  512.329200          NaN      NaN

In [2]:
train_data.dtypes

Survived      int64
Pclass        int64
Name         object
Sex          object
Age         float64
SibSp         int64
Parch         int64
Ticket       object
Fare        float64
Cabin        object
Embarked     object
dtype: object

The describe( ) method gives us a quick summary of what we are dealing with. For example, we know that there are a total of 11 variables (excluding passenger ID), and there are a mixture of variables which are categorical ('Ticket', 'Cabin', 'Embarked') as well as numerical ('Pclass', 'Age', 'Fare').

Let's take a look at the first 5 rows of the dataset. We can subset the data using square brackets, [ ], or use the DataFrame method, `head()`.

In [3]:
train_data[:10]

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
10                  1       2   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
6                0      0            330877   8.4583   NaN        Q  
7                0      0             17463  51.8625   E46        S  
8                3      1            349909  21.0750   NaN        S  
9                0      2            347742  11.1333   NaN        S  
10               1      0            237736  30.0708   NaN        C

In [4]:
train_data.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

### Subsetting Data

Now that we have a better idea of what the data looks like, we can conduct some data analysis. Suppose we are interested in finding out who embarked at location, 'S'. One way to do so is by writing the following code, and it returns all passengers who embarked at the location, 'S'.

In the following code block, we have provided a condition `[train_data['Embarked'] == 'S'` - that the column, "Embarked" should be equivalent to "S". The `[10]` refers to the first 10 rows that satisfies this condition.

In [5]:
# We can pass in conditions to subset the data
train_data[train_data['Embarked'] == 'S']

Survived  Pclass  \
PassengerId                     
1                   0       3   
3                   1       3   
4                   1       1   
5                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
11                  1       3   
12                  1       1   
13                  0       3   
14                  0       3   
15                  0       3   
16                  1       2   
18                  1       2   
19                  0       3   
21                  0       2   
22                  1       2   
24                  1       1   
25                  0       3   
26                  1       3   
28                  0       1   
30                  0       3   
34                  0       2   
36                  0       1   
38                  0       3   
39                  0       3   
41                  0       3   
42                  0       2   
46                  0       3   
50                  0       3   
...               ...     ...   
852                 0       3   
854                 1       1   
855                 0       2   
856                 1       3   
857                 1       1   
858                 1       1   
861                 0       3   
862                 0       2   
863                 1       1   
864                 0       3   
865                 0       2   
866                 1       2   
868                 0       1   
869                 0       3   
870                 1       3   
871                 0       3   
872                 1       1   
873                 0       1   
874                 0       3   
877                 0       3   
878                 0       3   
879                 0       3   
881                 1       2   
882                 0       3   
883                 0       3   
884                 0       2   
885                 0       3   
887                 0       2   
888                 1       1   
889                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
7                                      McCarthy, Mr. Timothy J    male  54.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
11                             Sandstrom, Miss. Marguerite Rut  female   4.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
13                              Saundercock, Mr. William Henry    male  20.0   
14                                 Andersson, Mr. Anders Johan    male  39.0   
15                        Vestrom, Miss. Hulda Amanda Adolfina  female  14.0   
16                            Hewlett, Mrs. (Mary D Kingcome)   female  55.0   
18                                Williams, Mr. Charles Eugene    male   NaN   
19           Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0   
21                                        Fynney, Mr. Joseph J    male  35.0   
22                                       Beesley, Mr. Lawrence    male  34.0   
24                                Sloper, Mr. William Thompson    male  28.0   
25                               Palsson, Miss. Torborg Danira  female   8.0   
26           Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  female  38.0   
28                              Fortune, Mr. Charles Alexander    male  19.0   
30                                         Todoroff, Mr. Lalio    male   NaN   
34           

Other than subsetting data on only 1 conditions, we can also subset data on 2 or more conditions, using the ampersand (`&`) logo.

In [6]:
train_data[(train_data['Embarked'] == 'S') & (train_data['Pclass'] == 1) & (train_data['Age'] > 30)]

Survived  Pclass  \
PassengerId                     
4                   1       1   
7                   0       1   
12                  1       1   
36                  0       1   
63                  0       1   
93                  0       1   
111                 0       1   
125                 0       1   
138                 0       1   
171                 0       1   
188                 1       1   
225                 1       1   
231                 1       1   
249                 1       1   
253                 0       1   
263                 0       1   
264                 0       1   
269                 1       1   
270                 1       1   
276                 1       1   
319                 1       1   
332                 0       1   
333                 0       1   
340                 0       1   
384                 1       1   
391                 1       1   
435                 0       1   
439                 0       1   
448                 1       1   
450                 1       1   
...               ...     ...   
578                 1       1   
610                 1       1   
622                 1       1   
626                 0       1   
631                 1       1   
661                 1       1   
663                 0       1   
672                 0       1   
691                 1       1   
695                 0       1   
702                 1       1   
708                 1       1   
713                 1       1   
742                 0       1   
746                 0       1   
760                 1       1   
764                 1       1   
766                 1       1   
780                 1       1   
797                 1       1   
807                 0       1   
810                 1       1   
821                 1       1   
823                 0       1   
857                 1       1   
858                 1       1   
863                 1       1   
868                 0       1   
872                 1       1   
873                 0       1   

                                                          Name     Sex   Age  \
PassengerId                                                                    
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
7                                      McCarthy, Mr. Timothy J    male  54.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
36                              Holverson, Mr. Alexander Oskar    male  42.0   
63                                 Harris, Mr. Henry Birkhardt    male  45.0   
93                                 Chaffee, Mr. Herbert Fuller    male  46.0   
111                             Porter, Mr. Walter Chamberlain    male  47.0   
125                                White, Mr. Percival Wayland    male  54.0   
138                                Futrelle, Mr. Jacques Heath    male  37.0   
171                                  Van der hoef, Mr. Wyckoff    male  61.0   
188              Romaine, Mr. Charles Hallace ("Mr C Rolmane")    male  45.0   
225                               Hoyt, Mr. Frederick Maxfield    male  38.0   
231               Harris, Mrs. Henry Birkhardt (Irene Wallach)  female  35.0   
249                              Beckwith, Mr. Richard Leonard    male  37.0   
253                                  Stead, Mr. William Thomas    male  62.0   
263                                          Taussig, Mr. Emil    male  52.0   
264                                      Harrison, Mr. William    male  40.0   
269              Graham, Mrs. William Thompson (Edith Junkins)  female  58.0   
270                                     Bissette, Miss. Amelia  female  35.0   
276                          Andrews, Miss. Kornelia Theodosia  female  63.0   
319                                   Wick, Miss. Mary Natalie  female  31.0   
332                                        Partner, Mr. Austen    male  45.5   
333          

### Summary Statistics

After subsetting data, we may be interested in finding the mean or variance of the observations that satisfy a particular condition. One can call the `.mean()` method to find the mean of a particular variable of interest, when the `np.var()` function can be used to calculate the variance of the variable of interest.

In [7]:
train_data.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [8]:
train_data[train_data['Fare'] > 100].describe()

Survived  Pclass        Age      SibSp      Parch        Fare
count  53.000000    53.0  48.000000  53.000000  53.000000   53.000000
mean    0.735849     1.0  31.956667   0.660377   0.886792  187.032858
std     0.445099     0.0  14.550628   0.830741   0.973897   96.462925
min     0.000000     1.0   0.920000   0.000000   0.000000  106.425000
25%     0.000000     1.0  21.750000   0.000000   0.000000  120.000000
50%     1.000000     1.0  33.000000   0.000000   1.000000  151.550000
75%     1.000000     1.0  40.250000   1.000000   2.000000  227.525000
max     1.000000     1.0  64.000000   3.000000   4.000000  512.329200

If we want to find the variance of a specific column, we can use the function `np.var`. Similarly, we can find the average fare using the function, `np.mean(train_data['Fare'])`. There exists other functions too, such as `np.median`.

In [9]:
np.var(train_data[(train_data['Embarked'] == 'S') & (train_data['Pclass'] == 1)]['Fare'])

3431.532010935895

In [10]:
print('Average fare is', np.mean(train_data['Fare']))
print('Fare variance is', np.var(train_data['Fare']))

Average fare is 32.2042079685746
Fare variance is 2466.6653116850434


In-class assignment:

1. Previously, we note that the mean survival rate of the dataset is 38.4% (where did we note this?). Suppose you think that passengers who embarked at 'S' has a higher rate of surviving. How do you check this?
2. Are males or females more likely to perish?
3. Which passenger class costs the most?

In [11]:
### Your code here
train_data[train_data['Embarked'] == 'S'].describe()['Survived']
np.mean(train_data[train_data['Embarked'] == 'S']['Survived'])

0.33695652173913043

In [12]:
train_data[train_data['Sex'] == 'male'].describe()['Survived']
np.mean(train_data[train_data['Sex'] == 'male']['Survived'])

0.18890814558058924

In [13]:
train_data[train_data['Sex'] == 'female'].describe()['Survived']
np.mean(train_data[train_data['Sex'] == 'female']['Survived'])

0.7420382165605095

In [14]:
train_data[train_data['Pclass'] == 1].describe()['Fare']
np.mean(train_data[train_data['Pclass'] == 1]['Fare'])

84.15468749999992

In [15]:
train_data[train_data['Pclass'] == 2].describe()['Fare']
np.mean(train_data[train_data['Pclass'] == 2]['Fare'])

20.66218315217391

In [16]:
train_data[train_data['Pclass'] == 3].describe()['Fare']
np.mean(train_data[train_data['Pclass'] == 3]['Fare'])

13.675550101832997

### Data Manipulation
Now that we have a basic understanding of data manipulation in Python, let's move on to more interesting stuff:

1. Locating and imputing missing values in the data
2. Pivot Tables
3. Preliminary Text Cleaning
4. Groupby

---

### 1. Locating and Imputing Missing Values
To identify all missing values in the dataset, we can use the function `pd.isnull()`, and pass in our training data as the argument.

In [17]:
pd.isnull(train_data).head()

Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
PassengerId                                                                
1               False   False  False  False  False  False  False   False   
2               False   False  False  False  False  False  False   False   
3               False   False  False  False  False  False  False   False   
4               False   False  False  False  False  False  False   False   
5               False   False  False  False  False  False  False   False   

              Fare  Cabin  Embarked  
PassengerId                          
1            False   True     False  
2            False  False     False  
3            False   True     False  
4            False  False     False  
5            False   True     False

To obtain the total number of missing values for each column, we can use the function, `np.sum()`. Note that `np.sum()` sums across each column. To sum across each row, we can pass in an additional argument, `axis=1`.

In [18]:
#otherwise, you can use the command, ?np.sum(), to check the arguments used in the function

In [19]:
np.sum(pd.isnull(train_data))

Survived      0
Pclass        0
Name          0
Sex           0
Age         177
SibSp         0
Parch         0
Ticket        0
Fare          0
Cabin       687
Embarked      2
dtype: int64

To talk about the data types when importing csv files in Python

Data Cleaning:

It appears that many of the Age and Cabin data are missing. In this case, we can do the following, depending on the column category:

1. Fill the mean value (with `df.fill_na()`)
2. Create a new category
3. Drop the columns

---

The DataFrame method `fillna()`, fills all missing values with an integer, float, or string provided by the user. Setting the argument, `inplace=True` results in the dataframe being altered. Otherwise, you can leave out the argument `inplace=True`.

In [20]:
train_data['Age'].fillna(np.mean(train_data['Age']), inplace=True)

In [21]:
# Filling in mean value for Age
def fill_mean_value(train_data, column): 
    train_data[column].fillna(np.mean(train_data[column]), inplace=True)
    return train_data

new_train_data = fill_mean_value(train_data, 'Age')

Next, we note that there were many missing values for the variable, "Cabin". One may posit that these individuals may not have been assigned to any cabins. We can create a new category, by filling the missing values with a particular string, "No Cabin".

In [22]:
np.sum(pd.isnull(train_data))

Survived      0
Pclass        0
Name          0
Sex           0
Age           0
SibSp         0
Parch         0
Ticket        0
Fare          0
Cabin       687
Embarked      2
dtype: int64

In [23]:
# Creating a new category
def add_new_category(train_data, column, new_cat_name):
    train_data[column].fillna(new_cat_name, inplace=True)
    return train_data

new_train_data = add_new_category(new_train_data, 'Cabin', 'No Cabin')
new_train_data.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare     Cabin Embarked  
PassengerId                                                              
1                1      0         A/5 21171   7.2500  No Cabin        S  
2                1      0          PC 17599  71.2833       C85        C  
3                0      0  STON/O2. 3101282   7.9250  No Cabin        S  
4                1      0            113803  53.1000      C123        S  
5                0      0            373450   8.0500  No Cabin        S

In [24]:
new_train_data.pivot_table(values='Survived', 
                           index='SibSp',
                           columns='Parch',
                           aggfunc='mean')

Parch         0         1         2         3    4         5    6
SibSp                                                            
0      0.303538  0.657895  0.724138  1.000000  0.0  0.000000  NaN
1      0.520325  0.596491  0.631579  0.333333  0.0  0.333333  0.0
2      0.250000  0.857143  0.500000  1.000000  NaN       NaN  NaN
3      1.000000  0.000000  0.285714       NaN  NaN       NaN  NaN
4           NaN  0.000000  0.333333       NaN  NaN       NaN  NaN
5           NaN       NaN  0.000000       NaN  NaN       NaN  NaN
8           NaN       NaN  0.000000       NaN  NaN       NaN  NaN

In [25]:
# Creates a new column called CaBinary, such that if entry is No Cabin, returns 0
new_train_data['CaBinary'] = new_train_data['Cabin'] != 'No Cabin'
new_train_data.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare     Cabin Embarked  \
PassengerId                                                               
1                1      0         A/5 21171   7.2500  No Cabin        S   
2                1      0          PC 17599  71.2833       C85        C   
3                0      0  STON/O2. 3101282   7.9250  No Cabin        S   
4                1      0            113803  53.1000      C123        S   
5                0      0            373450   8.0500  No Cabin        S   

             CaBinary  
PassengerId            
1               False  
2                True  
3               False  
4                True  
5               False

Alternatively, for columns with too many missing values, one can also opt to drop these columns entirely, since they may not be too informative for prediction. 

In [26]:
del new_train_data['CaBinary']

In [27]:
def drop_columns(train_data, columns_to_drop):
    '''
    columns_to_drop is a list
    '''
    columns_to_keep = []
    
    # The following for loop keeps columns that I want
    for column in train_data.columns:
        if column not in columns_to_drop:
            columns_to_keep.append(column)

    return train_data[columns_to_keep]

new_train_data = drop_columns(new_train_data, ['Cabin'])
new_train_data.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Embarked  
PassengerId                                                    
1                1      0         A/5 21171   7.2500        S  
2                1      0          PC 17599  71.2833        C  
3                0      0  STON/O2. 3101282   7.9250        S  
4                1      0            113803  53.1000        S  
5                0      0            373450   8.0500        S

### 2. Pivot Tables

We can use the method, `pivot_table` to generate conditional summary statistics. For example, one might be interested in the number of male or female passengers who survived, and embarked at different locations. 

1. Gender of passengers
2. Embarkation point
3. Survival Rate

In that case, one can use the following code:

In [28]:
# aggfunc set to mean, because I'm interested in average survival rate
# this is conditional on gender and embarkation point
new_train_data.pivot_table(values='Survived', 
                           index='Sex', 
                           columns='Embarked',
                           aggfunc='mean')

# For pivot tables, we want to condition on categorical columns (at most a few categories)

Embarked         C         Q         S
Sex                                   
female    0.876712  0.750000  0.689655
male      0.305263  0.073171  0.174603

Here, we created a pivot table, with the arguments `values='Survived`, `index='Sex'`, `columns='Embarked'` and `aggfunc='mean'`. This means that we want to find the mean survival rates, by gender and embarkation point. 

In-class assignment:

1. Using the DataFrame method, `pivot_table`, find the average fare paid by female passengers who survived. - "Fare", "Sex", "Survived"


2. Using the DataFrame method, `pivot_table`, find the average age of male passengers who were of passenger class 1. - "Age", "Sex", "Pclass"

In [29]:
## Your code here
new_train_data.pivot_table(values='Fare', 
                           index='Sex', 
                           columns='Survived',
                           aggfunc='mean')

Survived          0          1
Sex                           
female    23.024385  51.938573
male      21.960993  40.821484

In [30]:
## Your code here
new_train_data.pivot_table(values='Age', 
                           index='Sex', 
                           columns=['Pclass', 'Survived'],
                           aggfunc='mean')

Pclass            1                     2                     3           
Survived          0          1          0          1          0          1
Sex                                                                       
female    25.666667  34.420792  36.000000  28.127118  25.206736  22.930249
male      41.489427  35.520346  33.086745  17.631073  27.948083  23.696001

### 3. Preliminary Text Analysis

When provided with textual data, we can conduct text analysis on the data. Sometimes, we want to extract certain information from a text, for example, we may be interested in extracting the email addresses in an email.

As a start, let's take a closer look at the "Ticket" column of the dataset.

In [31]:
x = "Hello World"
?x.split()

In [32]:
new_train_data['Name']

PassengerId
1                                Braund, Mr. Owen Harris
2      Cumings, Mrs. John Bradley (Florence Briggs Th...
3                                 Heikkinen, Miss. Laina
4           Futrelle, Mrs. Jacques Heath (Lily May Peel)
5                               Allen, Mr. William Henry
6                                       Moran, Mr. James
7                                McCarthy, Mr. Timothy J
8                         Palsson, Master. Gosta Leonard
9      Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
10                   Nasser, Mrs. Nicholas (Adele Achem)
11                       Sandstrom, Miss. Marguerite Rut
12                              Bonnell, Miss. Elizabeth
13                        Saundercock, Mr. William Henry
14                           Andersson, Mr. Anders Johan
15                  Vestrom, Miss. Hulda Amanda Adolfina
16                      Hewlett, Mrs. (Mary D Kingcome) 
17                                  Rice, Master. Eugene
18                 

In [33]:
new_train_data['Ticket'].head()

PassengerId
1           A/5 21171
2            PC 17599
3    STON/O2. 3101282
4              113803
5              373450
Name: Ticket, dtype: object

One may posit that the numbers on the ticket numbers may be informative of the ticket type of the passengers. For example, one can test the hypothesis that a larger ticket number may be more expensive than a cheaper ticket. Before we can test this hypothesis, we need to extract the "numeric" part of the ticket.

To do so, we first note that the "character" part of the ticket and the "numeric" part of the ticket are split by a space. Thus, we can split each ticket into 2 parts by applying the "split" function on a string to obtain a list of length 2.

In [34]:
new_train = new_train_data['Ticket'].loc[1]
new_train.split(' ')

['A/5', '21171']

##### Functional/Lambda Programming

Note that the ticket column contains both integers and characters. We can define a simple function that separates on the space. This is a good time to introduce lambda programming. A lambda function is an anonymous one - it does not have a "name". Typically, programmers use lambda functions when they are intending to use the function once.


Below, we show an example of a lambda function. Here, the argument for the lambda function is `x`, and the lambda function basically splits the argument, `x` by a space.

In [35]:
# Split on ", " for each name in the column "Name"
# First i obtain a list, then i extract last element (a string)
# Then i split on the string to obtain another list
# Obtain first element of this list and name column, "Title"
new_train_data['Title'] = new_train_data['Name'].apply(lambda x: (x.split(', ')[-1]).split('. ')[0])

Suppose we are only interested in the integer component of the ticket code. How can we do so? It turns out that to extract the last term in a list, we can subset the list using `[-1]`.

In [36]:
['a', 'b', 'c'][-1]

'c'

In [37]:
new_train_data['Ticket'].apply(lambda x: x.split(' ')[-1]).head()

PassengerId
1      21171
2      17599
3    3101282
4     113803
5     373450
Name: Ticket, dtype: object

What if, instead, we are interested in the string component of the ticket code? How can we get it? One way of deriving the string component of the code is through the following:

In [38]:
def get_string_component_from_ticket(data, column='Ticket'):
    string_lst = []
    for row in new_train_data['Ticket'].apply(lambda x: x.split(' ')):
        if len(row) > 1: string_lst.append(row[0]) # if there are more than 2 args, 1st arg is the string component
        else: string_lst.append('Null') # otherwise, there are no string components
    return string_lst
        
get_string_component_from_ticket(new_train_data)[:5]

['A/5', 'PC', 'STON/O2.', 'Null', 'Null']

Using the previous 2 chunks of codes, we can dissect the Ticket column into 2 components: the string component and the integer component. Let's write some code to do that:

In [39]:
def dissect_ticket_column(data, column='Ticket'):
    int_component = data[column].apply(lambda x: x.split(' ')[-1])
    str_component = get_string_component_from_ticket(new_train_data)
    
    data['Ticket_Int'] = int_component # creates a new column that stores the integer component
    data['Ticket_Str'] = str_component # creates a new column that stores the string component
    
    return data

new_data = dissect_ticket_column(new_train_data)
new_data.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Embarked Title  \
PassengerId                                                           
1                1      0         A/5 21171   7.2500        S    Mr   
2                1      0          PC 17599  71.2833        C   Mrs   
3                0      0  STON/O2. 3101282   7.9250        S  Miss   
4                1      0            113803  53.1000        S   Mrs   
5                0      0            373450   8.0500        S    Mr   

            Ticket_Int Ticket_Str  
PassengerId                        
1                21171        A/5  
2                17599         PC  
3              3101282   STON/O2.  
4               113803       Null  
5               373450       Null

Looks like everything worked out fine! We have successfully split the ticket column into 2 different columns for us to carry out analysis on. It might be useful to conduct a similar analysis on the column, Name.

---

In-class assignment (You can define your own functions, or use functional programming):
1. Split the name column of new_data in to 2 different columns, Last_Name, and First_Name. Hint: You can split on the comma punctuation (', ') - there's a space.  
2. After splitting, do another split on the First_Name column on the following string: ". " (note the space) to obtain the honorific name (Mr, Ms etc.) and the actual Last_Name. Name the columns: Honorific and True_Last_Name.

In [40]:
int()

0

In [41]:
def dissect_name_column(data, column='Name'):
    ## Your code here
    last_name = data[column].apply(lambda x: x.split(', ')[0]) # Split Name column by ', ' and take element 0
    first_name = data[column].apply(lambda x: x.split(', ')[1]) # Split Name column by ', ' and take element 1
    
    title = first_name.apply(lambda x: x.split('. ')[0]) # Split first_name column by '. ' and take element 0
    
    data['True_Last_Name'] = last_name
    data['Honorific'] = title
    return data

new_train_data = dissect_name_column(data=new_train_data, column='Name')

In [42]:
np.unique(new_train_data['Honorific'])

array(['Capt', 'Col', 'Don', 'Dr', 'Jonkheer', 'Lady', 'Major', 'Master',
       'Miss', 'Mlle', 'Mme', 'Mr', 'Mrs', 'Ms', 'Rev', 'Sir',
       'the Countess'], dtype=object)

In [43]:
new_train_data.pivot_table(values='Survived',
                           index='Title',
                           aggfunc='mean')

Survived
Title                 
Capt          0.000000
Col           0.500000
Don           0.000000
Dr            0.428571
Jonkheer      0.000000
Lady          1.000000
Major         0.500000
Master        0.575000
Miss          0.697802
Mlle          1.000000
Mme           1.000000
Mr            0.156673
Mrs           0.792000
Ms            1.000000
Rev           0.000000
Sir           1.000000
the Countess  1.000000

Going forward, it is useful to check what kinds of tickets (Ticket_Str) there are. To do so, we can use the `np.unique` function to do that (what other functions could we have used?). Let's do that.

In [44]:
np.unique(new_data['Ticket_Str'])

array(['A./5.', 'A.5.', 'A/4', 'A/4.', 'A/5', 'A/5.', 'A/S', 'A4.', 'C',
       'C.A.', 'C.A./SOTON', 'CA', 'CA.', 'F.C.', 'F.C.C.', 'Fa', 'Null',
       'P/PP', 'PC', 'PP', 'S.C./A.4.', 'S.C./PARIS', 'S.O./P.P.',
       'S.O.C.', 'S.O.P.', 'S.P.', 'S.W./PP', 'SC', 'SC/AH', 'SC/PARIS',
       'SC/Paris', 'SCO/W', 'SO/C', 'SOTON/O.Q.', 'SOTON/O2', 'SOTON/OQ',
       'STON/O', 'STON/O2.', 'SW/PP', 'W./C.', 'W.E.P.', 'W/C', 'WE/P'],
      dtype=object)

It appears that the data is slightly dirty - there appears to be many mistakes (we actually do not know if these are genuine errors or not). One way to verify this is to count the number of times each category appears. If a category only appears once, and is lexically close to another category, then it is very likely to be a recording error.

One way to do so is through the use of a dictionary, looping over every value and keeping count of them. Alternatively, we can use the Series method (a Series is basically a column of a DataFrame), `value_counts`. A Series is basically a column of the DataFrame.

In [45]:
new_data['Ticket_Str'].value_counts()

Null          665
PC             60
C.A.           27
STON/O         12
A/5            10
W./C.           9
CA.             8
SOTON/O.Q.      8
SOTON/OQ        7
A/5.            7
CA              6
STON/O2.        6
F.C.C.          5
C               5
SC/PARIS        5
S.O.C.          5
SC/Paris        4
PP              3
S.O./P.P.       3
SC/AH           3
A/4.            3
A/4             3
SOTON/O2        2
WE/P            2
S.C./PARIS      2
P/PP            2
A./5.           2
A.5.            2
F.C.            1
C.A./SOTON      1
SCO/W           1
S.W./PP         1
SW/PP           1
SO/C            1
S.C./A.4.       1
W.E.P.          1
A4.             1
S.O.P.          1
A/S             1
Fa              1
S.P.            1
W/C             1
SC              1
Name: Ticket_Str, dtype: int64

In-class Assignment:

Use the dictionary method to generate the same table above. Helper code has been provided below.

In [46]:
def unique_values_dict(data, column='Ticket_Str'):
    ticket_dict = {}
    for ticket_value in new_data[column].values:
        # Your code here
        if ticket_value in ticket_dict.keys():
            ticket_dict[ticket_value] += 1
        else:
            ticket_dict[ticket_value] = 1
    return ticket_dict

unique_values_dict(new_data, "Ticket_Str")

{'A./5.': 2,
 'A.5.': 2,
 'A/4': 3,
 'A/4.': 3,
 'A/5': 10,
 'A/5.': 7,
 'A/S': 1,
 'A4.': 1,
 'C': 5,
 'C.A.': 27,
 'C.A./SOTON': 1,
 'CA': 6,
 'CA.': 8,
 'F.C.': 1,
 'F.C.C.': 5,
 'Fa': 1,
 'Null': 665,
 'P/PP': 2,
 'PC': 60,
 'PP': 3,
 'S.C./A.4.': 1,
 'S.C./PARIS': 2,
 'S.O./P.P.': 3,
 'S.O.C.': 5,
 'S.O.P.': 1,
 'S.P.': 1,
 'S.W./PP': 1,
 'SC': 1,
 'SC/AH': 3,
 'SC/PARIS': 5,
 'SC/Paris': 4,
 'SCO/W': 1,
 'SO/C': 1,
 'SOTON/O.Q.': 8,
 'SOTON/O2': 2,
 'SOTON/OQ': 7,
 'STON/O': 12,
 'STON/O2.': 6,
 'SW/PP': 1,
 'W./C.': 9,
 'W.E.P.': 1,
 'W/C': 1,
 'WE/P': 2}

There appears to be quite a number of mistakes, based on our analysis. For example, A./5., A.5., A/5 and A/5. seem to be referring to the same cabin type. Same goes for the types A/4 and A/4 and A/4.. To clean the data, we can write a simple function that removes all kinds of punctuations in the data.

Let's search for ["clean punctuations python"](https://www.google.com/search?q=clean+punctuations+python&oq=clean+punctuations+python&aqs=chrome..69i57j0l4.2954j0j1&sourceid=chrome&ie=UTF-8) using Google's search engine.

Google recommends using the library string, and module punctuation. Let's find out what's in `string.punctuation first'.

In [47]:
import string

string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

The following code was taken from [StackOverflow](https://stackoverflow.com/questions/265960/best-way-to-strip-punctuation-from-a-string). Essentially, the code takes a data structure as its input, loops over each element (each element here refers to a string) in the data structure, and removes all punctuation from the string.

In [48]:
def clean_punctuations(data, column):
    clean_data = []
    for idx, text in enumerate(data[column].values):
        interim_string = text
        for punctuation in string.punctuation:
            if punctuation in text:
                interim_string = interim_string.replace(punctuation, '')
        # clean_data.append(row.translate(str.maketrans('', '', string.punctuation)))
        clean_data.append(interim_string)
    return clean_data

clean_punctuations(new_data, 'Ticket_Str')[0]

'A5'

All punctuation marks have been removed from the `Ticket_Str` column! This is a neat trick that you can use on any specific datasets to remove puncutations or any other character that may be uninformative or redundant. Now that we have cleaned up the dataset, let's take a look at the remaining unique values in the data.

In [49]:
print(len(np.unique(new_data['Ticket_Str'])))
print(len(np.unique(clean_punctuations(new_data, 'Ticket_Str'))))

43
30


In [50]:
# Convert to lower case
new_data['New_Ticket_Str'] = clean_punctuations(new_data, 'Ticket_Str')
new_data['New_Ticket_Str'].apply(lambda x: x.lower())
new_data['New_Ticket_Str'] = new_data['New_Ticket_Str'].apply(lambda x: x.lower())

In [51]:
new_data['New_Ticket_Str'].head()

PassengerId
1        a5
2        pc
3    stono2
4      null
5      null
Name: New_Ticket_Str, dtype: object

#### Data Imputation for Categorical Variables

Let's take a look at the missing values in the data once again. Note that we have 2 observations with missing values.

In [52]:
np.sum(pd.isnull(new_train_data))

Survived          0
Pclass            0
Name              0
Sex               0
Age               0
SibSp             0
Parch             0
Ticket            0
Fare              0
Embarked          2
Title             0
Ticket_Int        0
Ticket_Str        0
True_Last_Name    0
Honorific         0
New_Ticket_Str    0
dtype: int64

In [53]:
# I subset on observations where Embarked column is null
new_train_data[pd.isnull(new_train_data['Embarked'])]

Survived  Pclass                                       Name  \
PassengerId                                                                
62                  1       1                        Icard, Miss. Amelie   
830                 1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

                Sex   Age  SibSp  Parch  Ticket  Fare Embarked Title  \
PassengerId                                                            
62           female  38.0      0      0  113572  80.0      NaN  Miss   
830          female  62.0      0      0  113572  80.0      NaN   Mrs   

            Ticket_Int Ticket_Str True_Last_Name Honorific New_Ticket_Str  
PassengerId                                                                
62              113572       Null          Icard      Miss           null  
830             113572       Null          Stone       Mrs           null

In [54]:
# Suppose I believe that I can embarkation point is related to fare
new_train_data.pivot_table(values='Fare',
                           index='Pclass',
                           columns='Embarked',
                           aggfunc='mean')

Embarked           C          Q          S
Pclass                                    
1         104.718529  90.000000  70.364862
2          25.358335  12.350000  20.327439
3          11.214083  11.183393  14.644083

In [55]:
# Since 80 is closer to 70.36 than 90, I assign these 2 passengers "S"
new_train_data['Embarked'].fillna('S', inplace=True)

In [56]:
np.sum(pd.isnull(new_train_data))

Survived          0
Pclass            0
Name              0
Sex               0
Age               0
SibSp             0
Parch             0
Ticket            0
Fare              0
Embarked          0
Title             0
Ticket_Int        0
Ticket_Str        0
True_Last_Name    0
Honorific         0
New_Ticket_Str    0
dtype: int64

When we are faced with such a scenario, one thing we might do is to to input the most common value in the dataset. To do this, we can use the `describe` method on the dataframe once again.

In [57]:
new_data.describe(include='all')

Survived      Pclass                  Name   Sex         Age  \
count   891.000000  891.000000                   891   891  891.000000   
unique         NaN         NaN                   891     2         NaN   
top            NaN         NaN  Waelens, Mr. Achille  male         NaN   
freq           NaN         NaN                     1   577         NaN   
mean      0.383838    2.308642                   NaN   NaN   29.699118   
std       0.486592    0.836071                   NaN   NaN   13.002015   
min       0.000000    1.000000                   NaN   NaN    0.420000   
25%       0.000000    2.000000                   NaN   NaN   22.000000   
50%       0.000000    3.000000                   NaN   NaN   29.699118   
75%       1.000000    3.000000                   NaN   NaN   35.000000   
max       1.000000    3.000000                   NaN   NaN   80.000000   

             SibSp       Parch Ticket        Fare Embarked Title Ticket_Int  \
count   891.000000  891.000000    891  891.000000      891   891        891   
unique         NaN         NaN    681         NaN        3    17        679   
top            NaN         NaN   1601         NaN        S    Mr       2343   
freq           NaN         NaN      7         NaN      646   517          7   
mean      0.523008    0.381594    NaN   32.204208      NaN   NaN        NaN   
std       1.102743    0.806057    NaN   49.693429      NaN   NaN        NaN   
min       0.000000    0.000000    NaN    0.000000      NaN   NaN        NaN   
25%       0.000000    0.000000    NaN    7.910400      NaN   NaN        NaN   
50%       0.000000    0.000000    NaN   14.454200      NaN   NaN        NaN   
75%       1.000000    0.000000    NaN   31.000000      NaN   NaN        NaN   
max       8.000000    6.000000    NaN  512.329200      NaN   NaN        NaN   

       Ticket_Str True_Last_Name Honorific New_Ticket_Str  
count         891            891       891            891  
unique         43            667        17             29  
top          Null      Andersson        Mr           null  
freq          665              9       517            665  
mean          NaN            NaN       NaN            NaN  
std           NaN            NaN       NaN            NaN  
min           NaN            NaN       NaN            NaN  
25%           NaN            NaN       NaN            NaN  
50%           NaN            NaN       NaN            NaN  
75%           NaN            NaN       NaN            NaN  
max           NaN            NaN       NaN            NaN

For the Embarked column, there are 3 unique values, with "S" as the most common value, appearing 644 out of 889 times.

---

In-class assignment:

1. Define a function that takes data and column as its input, and returns the most common value. You can assume that the column provided is a categorical or string column.
2. Define another function that fills the NA value with the most common value (refer to first function), and returns the data.

In [58]:
## Your code here
def most_common_value(z, z1):
    # Your code here
    return z[z1].value_counts().index[0]

def fill_na_with_mcv(x, x1):
    # Your code here
    mcv = most_common_value(x, x1)
    
    return x[x1].fillna(mcv)

def 

fill_na_with_mcv(new_data, 'Embarked')

SyntaxError: invalid syntax (<ipython-input-58-392848044b39>, line 12)

In [59]:
np.sum(pd.isnull(new_train_data))

Survived          0
Pclass            0
Name              0
Sex               0
Age               0
SibSp             0
Parch             0
Ticket            0
Fare              0
Embarked          0
Title             0
Ticket_Int        0
Ticket_Str        0
True_Last_Name    0
Honorific         0
New_Ticket_Str    0
dtype: int64

#### Correlation Analysis

Finally, we can conduct some meaningful analysis. One question we might ask is what factors are highly correlated with the Survival factor? 

To answer this question, we can check the correlation across different features, using the `corr` method on `new_data`. It returns the correlation coefficient between any 2 features in the dataframe. Note that the correlation has to lie between -1 and 1, and the larger the magnitude of the coefficient, the more correlated the 2 features are.

In [60]:
new_train_data.corr()

Survived    Pclass       Age     SibSp     Parch      Fare
Survived  1.000000 -0.338481 -0.069809 -0.035322  0.081629  0.257307
Pclass   -0.338481  1.000000 -0.331339  0.083081  0.018443 -0.549500
Age      -0.069809 -0.331339  1.000000 -0.232625 -0.179191  0.091566
SibSp    -0.035322  0.083081 -0.232625  1.000000  0.414838  0.159651
Parch     0.081629  0.018443 -0.179191  0.414838  1.000000  0.216225
Fare      0.257307 -0.549500  0.091566  0.159651  0.216225  1.000000

From the correlation matrix, it appears that the features, Fare and Passenger Class are highly correlated with the Survival Probability of an individual. However, we also note that the features, Sex and Embarked do not show up in this matrix. This is due to the fact that they are categorical or ordinal factors. However, we can still use the `pivot_table` method to find out the likelihood of a person surviving given a particular gender.

Below, we use the method, `pivot_table` to illustrate the relationship between survival likelihood against gender, passenger class and embarkation point.

In [61]:
new_train_data.pivot_table(values='Survived', 
                           index='Pclass', 
                           columns='Sex', 
                           aggfunc='mean')

Sex       female      male
Pclass                    
1       0.968085  0.368852
2       0.921053  0.157407
3       0.500000  0.135447

In [62]:
new_data.pivot_table(values='Survived', index='Sex', columns='Embarked', aggfunc='mean')

Embarked         C         Q         S
Sex                                   
female    0.876712  0.750000  0.692683
male      0.305263  0.073171  0.174603

As it turns out, Females have a very high rate of survival. This might be due to many reasons (they might be younger, or they may have paid a higher fare for their tickets etc.). Let's explore this phenomenon in larger detail. We can use the `pivot_table` method to do so.

---

In-class assignment:
1. Test our hypothesis that females are younger than males on the Titanic on average.
2. Test our hypothesis that females pay more than males on average.

In [63]:
## Your code here
new_train_data.pivot_table(values='Fare', index='Sex', aggfunc='mean')

Fare
Sex              
female  44.479818
male    25.523893

### 4. Groupby
One other thing that we can do with DataFrames is the `groupby` method. From the [groupby article on pandas](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html), a groupby operation involves some combination of splitting the object, applying a function, and combining the results. This can be used to group large amounts of data and compute operations on these groups.

---

As motivation, let's take a closer look at the "Ticket" column of the DataFrame. There are 2 ways to subset a column: `new_data['Ticket']` and `new_data.Ticket`. We prefer the first method since it allows for "spaces" in the variable name.

In [64]:
len(np.unique(new_train_data['Title']))

17

In [65]:
len(new_train_data['Title'])

891

We note that some of the passengers shared tickets: while there were only 681 unique tickets, there were a total of 891 passengers. This begs the question: who are the passengers who shared these tickets, and how many tickets were shared? 

As it turns out, we can answer this question using the DataFrame method, `new_data.groupby`. In our case, we can create a new Series that returns 1 when the ticket is shared by only 1 passenger, and 2 if it is shared by 2 passengers etc.

In [66]:
# First, we want to create a new series that depends on Title
groupby_title = new_train_data.groupby('Title')

In [67]:
# First, we want to create a new series that depends on Title
groupby_title = new_train_data.groupby('Title')

# Next, we want to identify the number of ticket holders (count) that shared the same ticket, 
# using the column "Name" as a reference
count_title = groupby_title.transform('count')
count_title

Survived  Pclass  Name  Sex  Age  SibSp  Parch  Ticket  Fare  \
PassengerId                                                                 
1                 517     517   517  517  517    517    517     517   517   
2                 125     125   125  125  125    125    125     125   125   
3                 182     182   182  182  182    182    182     182   182   
4                 125     125   125  125  125    125    125     125   125   
5                 517     517   517  517  517    517    517     517   517   
6                 517     517   517  517  517    517    517     517   517   
7                 517     517   517  517  517    517    517     517   517   
8                  40      40    40   40   40     40     40      40    40   
9                 125     125   125  125  125    125    125     125   125   
10                125     125   125  125  125    125    125     125   125   
11                182     182   182  182  182    182    182     182   182   
12                182     182   182  182  182    182    182     182   182   
13                517     517   517  517  517    517    517     517   517   
14                517     517   517  517  517    517    517     517   517   
15                182     182   182  182  182    182    182     182   182   
16                125     125   125  125  125    125    125     125   125   
17                 40      40    40   40   40     40     40      40    40   
18                517     517   517  517  517    517    517     517   517   
19                125     125   125  125  125    125    125     125   125   
20                125     125   125  125  125    125    125     125   125   
21                517     517   517  517  517    517    517     517   517   
22                517     517   517  517  517    517    517     517   517   
23                182     182   182  182  182    182    182     182   182   
24                517     517   517  517  517    517    517     517   517   
25                182     182   182  182  182    182    182     182   182   
26                125     125   125  125  125    125    125     125   125   
27                517     517   517  517  517    517    517     517   517   
28                517     517   517  517  517    517    517     517   517   
29                182     182   182  182  182    182    182     182   182   
30                517     517   517  517  517    517    517     517   517   
...               ...     ...   ...  ...  ...    ...    ...     ...   ...   
862               517     517   517  517  517    517    517     517   517   
863               125     125   125  125  125    125    125     125   125   
864               182     182   182  182  182    182    182     182   182   
865               517     517   517  517  517    517    517     517   517   
866               125     125   125  125  125    125    125     125   125   
867               182     182   182  182  182    182    182     182   182   
868               517     517   517  517  517    517    517     517   517   
869               517     517   517  517  517    517    517     517   517   
870                40      40    40   40   40     40     40      40    40   
871               517     517   517  517  517    517    517     517   517   
872               125     125   125  125  125    125    125     125   125   
873               517     517   517  517  517    517    517     517   517   
874               517     517   517  517  517    517    517     517   517   
875               125     125   125  125  125    125    125     125   125   
876               182     182   182  182  182    182    182     182   182   
877               517     517   517  517  517    517    517     517   517   
878               517     517   517  517  517    517    517     517   517   
879               517     517   517  517  517    517    517     517   517   
880               125     125   125  125  125    125    125     125   125   
881      

It turns out that passenger 4 is sharing the ticket with someone else.

We can use the following code to find out who passenger 4 is sharing the ticket with:

In [68]:
ticket_number = new_data['Ticket'].iloc[3]
new_data[new_data['Ticket'] == ticket_number]

Survived  Pclass                                          Name  \
PassengerId                                                                   
4                   1       1  Futrelle, Mrs. Jacques Heath (Lily May Peel)   
138                 0       1                   Futrelle, Mr. Jacques Heath   

                Sex   Age  SibSp  Parch  Ticket  Fare Embarked Title  \
PassengerId                                                            
4            female  35.0      1      0  113803  53.1        S   Mrs   
138            male  37.0      1      0  113803  53.1        S    Mr   

            Ticket_Int Ticket_Str True_Last_Name Honorific New_Ticket_Str  
PassengerId                                                                
4               113803       Null       Futrelle       Mrs           null  
138             113803       Null       Futrelle        Mr           null

Now that you have a working understanding of the DataFrame method, `groupby`, use it to solve the following questions:

1. Several passengers are sharing a cabin with one another. Use the DataFrame method, `groupby` to create a Series that returns how many passengers are sharing the same cabin.
2. Use the "First Name" column to create a Series that returns the number of passengers that are sharing the same First Name.

In [69]:
# Question 1
## Your code here

## Alternative code - gives the same result

In [70]:
# Question 2
## Your code here

Finally, we can save the dataset using the method `.to_csv()`. 

In [71]:
new_data.to_csv('data/newdata.csv')

In [72]:
new_data.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Embarked Title  \
PassengerId                                                           
1                1      0         A/5 21171   7.2500        S    Mr   
2                1      0          PC 17599  71.2833        C   Mrs   
3                0      0  STON/O2. 3101282   7.9250        S  Miss   
4                1      0            113803  53.1000        S   Mrs   
5                0      0            373450   8.0500        S    Mr   

            Ticket_Int Ticket_Str True_Last_Name Honorific New_Ticket_Str  
PassengerId                                                                
1                21171        A/5         Braund        Mr             a5  
2                17599         PC        Cumings       Mrs             pc  
3              3101282   STON/O2.      Heikkinen      Miss         stono2  
4               113803       Null       Futrelle       Mrs           null  
5               373450       Null          Allen        Mr           null

### Data Manipulation with MovieLens

The following example is taken from the book, "Python for Data Analysis" written by Wes McKinney, the inventor of Pandas. If you're interested in buying the book, you can find it [here](https://www.bookdepository.com/Python-for-Data-Analysis-2e-Wes-McKinney/9781491957660?ref=grid-view&qid=1564376975506&sr=1-1).

If you've taken a closer look at the MovieLens dataset, you would have noticed that the extension for the dataset is ".dat". However, it turns out that we can still use the `pd.read_csv` function for this. Alternatively, we can use the `pd.read_table` function for this as well. pandas has the ability to read different kinds of data files, and ".dat" is one of them. 

The MovieLens dataset has 3 different tables: 

1. "unames", a table that contains the usernames, gender, age and occupation etc. of users
2. "rnames", a table that contains the ratings, movie ids and timestamp etc.
3. "mnames", a table that contains the movie titles, movie ids and genres of the movies

We can add an argument, `names` to the functions `pd.read_table` or `pd.read_csv` to tell Python which table we are interested in.

In [76]:
# Import user data table
unames = ['user_id', 'gender', 'age', 'occupation', 'zip']
users = pd.read_csv('data/movielens/ml-1m/users.dat', engine='python', sep='::', header=None, names=unames)

# Import ratings table
rnames = ['user_id', 'movie_id', 'rating', 'timestamp']
ratings = pd.read_csv('data/movielens/ml-1m/ratings.dat', engine='python', sep='::', header=None, names=rnames)

# Import mnames table
mnames = ['movie_id', 'title', 'genres']
movies = pd.read_csv('data/movielens/ml-1m/movies.dat', engine='python', sep='::', header=None, names=mnames)

It is almost canon to take a look at the headers of the dataframes you have read after you have imported them into Python, so that's what we'll do in the next few code blocks.

In [77]:
users.head()

user_id gender  age  occupation    zip
0        1      F    1          10  48067
1        2      M   56          16  70072
2        3      M   25          15  55117
3        4      M   45           7  02460
4        5      M   25          20  55455

In [78]:
ratings.head()

user_id  movie_id  rating  timestamp
0        1      1193       5  978300760
1        1       661       3  978302109
2        1       914       3  978301968
3        1      3408       4  978300275
4        1      2355       5  978824291

In [79]:
movies.head()

movie_id                               title                        genres
0         1                    Toy Story (1995)   Animation|Children's|Comedy
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy
2         3             Grumpier Old Men (1995)                Comedy|Romance
3         4            Waiting to Exhale (1995)                  Comedy|Drama
4         5  Father of the Bride Part II (1995)                        Comedy

A keen observer would have noted that there are some dataframes that we can combine using columns that we can merge on. For example, one can merge the users and rating dataframes using the columns, user_id. Alternatively, one can also merge the ratings and movies dataset using the column, movie_id. 

As it turns out, we can use the function, `pd.merge` from pandas to merge 2 different datasets. Below, we merge the users and ratings dataframes on a common column, 'user_id'.

In [81]:
dat = pd.merge(users, ratings, on='user_id')

Let's take a look at the merged dataframe. Following the "general" principle of data analysis, we take a look at the first 5 rows of the dataframe.

In [82]:
dat.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp
0        1      F    1          10  48067      1193       5  978300760
1        1      F    1          10  48067       661       3  978302109
2        1      F    1          10  48067       914       3  978301968
3        1      F    1          10  48067      3408       4  978300275
4        1      F    1          10  48067      2355       5  978824291

We can now merge this dataframe with the movies dataframe using the column, 'movie_id'. Let's do that.

In [83]:
dat = pd.merge(dat, movies, on='movie_id')

In [84]:
dat.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  
4  One Flew Over the Cuckoo's Nest (1975)  Drama

In [85]:
dat.shape

(1000209, 10)

We managed to combine the 3 dataframes into 1, using the function `pd.merge`. This function takes as argument 2 dataframes to combine together, as well as some other arguments.

In [86]:
?pd.merge

Here are some following questions we can ask:

1. What are the most popular titles?
2. Which user rated the most movies? 
3. Which users are likely to give a very high rating on the movies that they rate?
4. What areas are users most likely to stay at?
5. Do older users tend to watch different genres of movies from the younger users?

It turns out that we can answer these questions fairly easily with the use of pandas and Numpy. Below, we answer each question in a sequential manner.

In [91]:
# Group dataset using titles and count using .size()
# This returns a data column
title_count = dat.groupby('title').size()

title_count

title
$1,000,000 Duck (1971)                              37
'Night Mother (1986)                                70
'Til There Was You (1997)                           52
'burbs, The (1989)                                 303
...And Justice for All (1979)                      199
1-900 (1994)                                         2
10 Things I Hate About You (1999)                  700
101 Dalmatians (1961)                              565
101 Dalmatians (1996)                              364
12 Angry Men (1957)                                616
13th Warrior, The (1999)                           750
187 (1997)                                          55
2 Days in the Valley (1996)                        286
20 Dates (1998)                                    139
20,000 Leagues Under the Sea (1954)                575
200 Cigarettes (1999)                              181
2001: A Space Odyssey (1968)                      1716
2010 (1984)                                        470
24 7

In [94]:
# Popular Titles
def top_n_titles(n):
    # .size() returns the number of times a title appears in the dataset
    titles = dat.groupby('title').size()
    return titles.sort_values(ascending=False)[:n] # we sort the number of counts in a descending order

top_n_titles(5)

title
American Beauty (1999)                                   3428
Star Wars: Episode IV - A New Hope (1977)                2991
Star Wars: Episode V - The Empire Strikes Back (1980)    2990
Star Wars: Episode VI - Return of the Jedi (1983)        2883
Jurassic Park (1993)                                     2672
dtype: int64

It appears that most of the users are fans of the movies, American Beauty, Star Wars and Jurassic Park. Now, which users rated the most movies? We can define a function that returns these users.

In [95]:
def most_active_users(n):
    active_users = dat.groupby('user_id').size() # we count the number of times a user appears in the dataset
    return active_users.sort_values(ascending=False)[:n] 

most_active_users(10)

user_id
4169    2314
1680    1850
4277    1743
1941    1595
1181    1521
889     1518
3618    1344
2063    1323
1150    1302
1015    1286
dtype: int64

One may be interested in users who are very critical or lax in their reviews of movies. If a user has a high mean rating, then it could be that his reviews are less informative than someone who is more critical.

In [106]:
# Argument in groupby is column we are grouping by is "user_id"
# Column we are interested in performing an operation (mean/avg) is "ratings"
# Operation is "mean"
dat.groupby('user_id')['rating'].mean().head()

user_id
1    4.188679
2    3.713178
3    3.901961
4    4.190476
5    3.146465
Name: rating, dtype: float64

In [107]:
def user_average_ratings(n):
    # "user_id" are my groups
    # "rating" is my column of interest
    # "mean" is my function
    user_ratings = dat.groupby('user_id')['rating'].mean()
    return user_ratings.sort_values(ascending=False)[:n]

user_average_ratings(10)

user_id
283     4.962963
2339    4.956522
3324    4.904762
3902    4.890909
446     4.843137
447     4.837838
4649    4.818182
4634    4.813725
1131    4.796117
4925    4.761905
Name: rating, dtype: float64

It appears that users 4169 and 1680 rated the most movies out of all the users. In addition, it appears that user 283 and 2339 are very lenient with their ratings, as their average rating comes to about 4.95 and above.

In [112]:
def most_populous_area(n):
    user_zip = dat.groupby('zip').size()
    return user_zip.sort_values(ascending=False)[:n]

most_populous_area(10)

zip
94110    3802
60640    3430
98103    3204
95616    3079
02138    3019
55408    2787
48135    2725
97401    2663
10025    2632
10024    2594
dtype: int64

Checking the zip codes, it appears that most users stay in the area of San Francisco and Illinois. Let's find the average age of the movie reviewers.

In [113]:
dat['age'].mean()

29.73831369243828

One may hypothesize that perhaps there is a relationship between a user's age, and the movie genres that he/she may enjoy. 

On average, the age of a reviewer is about 30. Let's categories the older reviewers as 1.5 times the average age, 45.

In [117]:
# Create a binary variable, "older" such that if age > 45, returns True
dat['older'] = dat['age'] > 45
dat.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres  older  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  False  
1  One Flew Over the Cuckoo's Nest (1975)  Drama   True  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  False  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  False  
4  One Flew Over the Cuckoo's Nest (1975)  Drama   True

In the next code block, we subset the data by 2 different conditions, and then count the number of times a particular genre appears for each subset.

In [118]:
def most_popular_genres(n):
    old_df = dat[dat['older'] == True]
    young_df = dat[dat['older'] == False]
    
    old_genres = old_df.groupby('genres').size()
    young_genres = young_df.groupby('genres').size()
    
    top_old_genres = old_genres.sort_values(ascending=False)[:n]
    top_young_genres = young_genres.sort_values(ascending=False)[:n]
    
    return top_old_genres, top_young_genres

top_old_genres, top_young_genres = most_popular_genres(10)

In [146]:
top_old_genres

genres
Drama              15791
Comedy             10313
Comedy|Romance      4762
Comedy|Drama        4744
Drama|Romance       3800
Action|Thriller     2350
Drama|War           2329
Drama|Thriller      2132
Thriller            1974
Horror              1775
dtype: int64

In [163]:
    movies_groupby = dat.groupby('title').size()
    greater_n_ratings = np.where(dat.groupby('title').size() >= 250)
    
    greater_n_ratings

(array([   3,    6,    7, ..., 3697, 3702, 3705]),)

Consider the following problem:

The Academy Awards (Oscars) has contacted us to select the top films of each year according to the movie genre. We have been given the following criteria for selection:

1. For each category, we are interested in movies with 250 ratings and above.
2. For each selected film, our tolerance for 1-star ratings is 1%.


In the following code block, 4 functions have been defined to help us achieve our objective of selecting the top movie that fulfils both requirements:

1. `popular_movies` - This function takes 2 arguments, a dataframe and an integer, and returns a dataframe of movies with more than n ratings.
2. `good_review_movies` - This function takes 2 arguments, a dataframe and a floating point, and returns movies with number of 1-star ratings < x%.
3. `best_movies` - This function takes 2 arguments, a dataframe and the genre, and returns the top movie for the genre given the dataframe.
4. `oscar_winners` - This function takes 3 arguments, a dataframe, an integer and a floating point, and returns a dictionary with genres of movies as the keys and the movie names as the values.

In [137]:
def popular_movies(data, n=250):
    '''
    Input: 
        1. data, a pd.DataFrame
        2. n, an integer
    
    Output: movies, pd.DataFrame
    
    Returns movies with more than n ratings
    '''
    # Groupby movies data to get movies that have at least N ratings
    movies_groupby = data.groupby('title').size()
    greater_n_ratings = np.where(data.groupby('title').size() >= n)

    top_movies = list(movies_groupby.iloc[greater_n_ratings].index)
    
    # After getting these top movies, create a new column that returns True if the movie is a top movie
    data['top_movies'] = data.title.apply(lambda x: x in top_movies)
    return data[data.top_movies == True]

def good_reviews_movies(data, tolerance=0.01):
    '''
    Input: 
        1. data, a pd.DataFrame
        2. tolerance, a floating point
    
    Output: good_movies, a pd.DataFrame
    
    Returns movies with number of 1-star rating < 1%
    '''
    # Group the movie ratings by the number of ratings each movie has
    movie_ratings_count = data.groupby('title')['movie_id'].count()
    
    # Subset dataframe into users who rated 1-star
    poor_df = data[data.rating == 1]
    poor_ratings_count = poor_df.groupby('title')['movie_id'].count()
    
    # Check for the proportion of 1-star ratings across all ratings
    tolerance_dict = dict()
    for movie in movie_ratings_count.index:
        if movie in poor_ratings_count.index:
            # Values are proportion of 1* ratings
            tolerance_dict[movie] = poor_ratings_count.loc[movie]/movie_ratings_count.loc[movie]
        else:
            tolerance_dict[movie] = 0
    
    # Check if each movie passes tolerance rate by iterating over movies in tolerance_dict
    good_review_movies = []
    for movie in tolerance_dict.keys():
        if tolerance_dict[movie] < tolerance: good_review_movies.append(movie)
    
    data['good_movies'] = data.title.apply(lambda x: x in good_review_movies)
    return data[data['good_movies'] == True]

def best_movie(data, genre):
    '''
    Input:
        1. data, a pd.DataFrame
        2. genre, a string
        
    Output:
        1. top_movie, a string
    
    Returns the top movie given a genre.
    '''
    # Subsets data by genre and calculates average ratings
    subset = data[data.genres == genre]
    movie_ratings = subset.groupby('title')['rating'].mean()
    
    # Returns top movie based on ratings
    top_movie = movie_ratings.sort_values(ascending=False).index[0]
    return top_movie
    
def oscar_winners(data, n=250, tolerance=0.01):
    '''
    Input:
        1. data, a pd.DataFrame
        2. n, an integer for the minimum number of ratings
        3. tolerance rate, a floating point, for proportion of 1-star ratings
        
    Output: top_movies, a dicttionary, key-value pairs are genre: movies
    
    Returns the Oscar winners for each category.
    '''
    # Get ratings of top movies with at least n reviews and has less than tolerance level
    movies_with_n_ratings = popular_movies(data, n)
    movies_with_good_reviews = good_reviews_movies(movies_with_n_ratings, tolerance)
    
    # Generate the Oscar Winners (based on our conditions)
    top_movies = dict()
    for genre in np.unique(data.genres.values):
        try: 
            oscar = best_movie(movies_with_good_reviews, genre)
            top_movies[genre] = oscar
        except: pass
    return top_movies

In [138]:
oscar_winners(dat, n=250, tolerance=0.01)

(array([   3,    6,    7, ..., 3697, 3702, 3705]),)


/Users/ngbolin/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:52: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


{'Action': 'Goldfinger (1964)',
 'Action|Adventure': 'Raiders of the Lost Ark (1981)',
 'Action|Adventure|Animation': 'Princess Mononoke, The (Mononoke Hime) (1997)',
 'Action|Adventure|Animation|Horror|Sci-Fi': 'Heavy Metal (1981)',
 "Action|Adventure|Children's|Comedy": 'Inspector Gadget (1999)',
 "Action|Adventure|Children's|Sci-Fi": 'Super Mario Bros. (1993)',
 'Action|Adventure|Comedy': 'Three Musketeers, The (1993)',
 'Action|Adventure|Comedy|Crime': 'Midnight Run (1988)',
 'Action|Adventure|Comedy|Horror': 'Evil Dead II (Dead By Dawn) (1987)',
 'Action|Adventure|Comedy|Horror|Sci-Fi': 'Army of Darkness (1993)',
 'Action|Adventure|Comedy|Romance': 'Princess Bride, The (1987)',
 'Action|Adventure|Comedy|Sci-Fi': 'Men in Black (1997)',
 'Action|Adventure|Crime': 'Rumble in the Bronx (1995)',
 'Action|Adventure|Crime|Drama': 'Batman (1989)',
 'Action|Adventure|Drama': 'Ben-Hur (1959)',
 'Action|Adventure|Drama|Romance': 'First Knight (1995)',
 'Action|Adventure|Drama|Sci-Fi|War': 'S

In [139]:

dat.to_csv('Movies.csv')

After writing our helper functions, let's proceed to find the Oscar Winners for each category!

In [ ]:
oscar_winners = oscar_winners(dat)

Let's take a look at the Oscar winners!

In [ ]:
oscar_winners

Given the dictionary, we can answer the following questions:
1. Which movies won the most Oscars?
2. Which movie, out of all the Oscar Winners, had the highest rating?
3. How many movies won more than 1 Oscar?

In [ ]:
oscars_won = {}
for movie in oscar_winners.values():
    # Have I seen the movie before in the oscars_won dictionary?
    if movie in oscars_won.keys():
        # If so, add 1 to the number of oscars it has already won
        oscars_won[movie] += 1
    else:
        # If not, initialize the first entry in the dictionary with 1
        oscars_won[movie] = 1

oscars_won

import operator

max(oscars_won.items(), key=operator.itemgetter(1))[0]

In [ ]:
best_movie = 'NA'
best_rating = 0

subset = dat.groupby('title')['rating'].mean()

for movie in oscars_won.keys():
    movie_rating = subset[movie]
    if movie_rating > best_rating:
        best_movie = movie
        best_rating = movie_rating

best_movie, best_rating